ltm/inserted_context_agent.py

```diff
@@ -0,0 +1,577 @@
+
+        context = [make_system_message(prompt.format(user_message=user_message, keywords=self.defined_kws))]
+        while True:
+            try:
+                print("Keyword gen")
+                response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+
+                keywords = [k.lower() for k in sanitize_and_parse_json(response)]
+                break
+            except Exception as e:
+                print(e)
+                continue
+
+        # Update known list of keywords
+        for k in keywords:
+            if k not in self.defined_kws:
+                self.defined_kws.append(k)
+
+        print(f"Interaction keywords: {keywords}")
+        return keywords
+
+    def create_context(self, user_message, max_prompt_size, previous_interactions, cost_cb):
+
+        stamped_user_message = str(datetime.now()) + ": " + user_message
+        context = [make_system_message(self.system_message.format(scratchpad=repr(self.scratchpad))), make_user_message(stamped_user_message)]
+        relevant_memories = self.get_relevant_memories(user_message, cost_cb)
+
+        # Get interactions from the memories
+        full_interactions = []
+        for m in relevant_memories:
+            interaction = self.messages_from_memory(m)
+            if interaction not in full_interactions:
+                full_interactions.append(interaction)
+
+        for m in full_interactions:
+            if "trivia" in m[0].content:
+                colour_print("YELLOW", f"<*** trivia ***>")
+            else:
+                colour_print("YELLOW", f"{m[0].content}")
+
+        # Add the previous messages
+        final_idx = self.session.message_count - 1
+        while previous_interactions > 0 and final_idx > 0:
+
+            # Agent reply
+            context.insert(1, self.session.by_index(final_idx).as_llm_dict())
+            # User message
+            context.insert(1, self.session.by_index(final_idx-1).as_llm_dict())
+
+            final_idx -= 2
+            previous_interactions -= 1
+
+        # Add in memories up to the max prompt size
+        current_size = token_counter(self.model, messages=context)
+        shown_mems = 0
+        target_size = max_prompt_size - self.max_message_size
+
+        for interaction in full_interactions[::-1]:
+            user_message, assistant_message = interaction
+            future_size = current_size + token_counter(self.model, messages=[user_message.as_llm_dict(), assistant_message.as_llm_dict()])
+
+            # If this message is going to be too big, then skip it
+            if future_size > target_size or shown_mems == 100:
+                continue
```

Comment:
It looks like this should be a `break` statement.

Reply:
I can split this out. Since everything being inserted are memories, I think that more memories are better, so I will prefer to skip memories that are too big, but insert any that can be inserted afterwards. Otherwise if a huge memory is retrieved, then the insertion will stop when we hit that memory and will end up with a gap that could otherwise be filled.

Reply:
Ah, I see it now. I thought that once the condition was met it would be always true, but you're right. Nice solution.